### PR TITLE
Fix host debug view not showing services before start

### PR DESCRIPTION
**Pull request by JamesNK**
Hosted services are now displayed before the host is started.

image

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs [hunk 1]
```diff
@@ -459,7 +459,9 @@ internal sealed class HostDebugView(Host host)
             public IHostEnvironment Environment => host._hostEnvironment;
             public IHostApplicationLifetime ApplicationLifetime => host._applicationLifetime;
             public HostOptions Options => host._options;
-            public List<IHostedService> HostedServices => new List<IHostedService>(host._hostedServices ?? Enumerable.Empty<IHostedService>());
+            // _hostedServices is null until the host is started. Resolve services directly from DI if host has started yet.
```
Comment (JamesNK) [inline src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs:462]:
```suggestion
            // _hostedServices is null until the host is started. Resolve services directly from DI if the host has yet to start.
```

File: src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs [hunk 2]
```diff
@@ -459,7 +459,9 @@ internal sealed class HostDebugView(Host host)
             public IHostEnvironment Environment => host._hostEnvironment;
             public IHostApplicationLifetime ApplicationLifetime => host._applicationLifetime;
             public HostOptions Options => host._options;
-            public List<IHostedService> HostedServices => new List<IHostedService>(host._hostedServices ?? Enumerable.Empty<IHostedService>());
+            // _hostedServices is null until the host is started. Resolve services directly from DI if the host has yet to start.
+            // This is ok because IHostedServices are registered with singleton lifetime so the same objects are returned.
```
Comment (steveharter) [inline src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs:463]: Could this cause side-affects around IHostedServices that use constructor injection?

Comment (JamesNK) [reply src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs:463]: Such as?

The only side effect I can think of is resolving services twice, but as I mentioned in the comment, `IHostedServices` are registered with a singleton lifetime.

Comment (steveharter) [reply src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs:463]: If a IHostedService implementation type is registered as a service (in addition to being registered as a hosted service) and it contains constructor args for other service(s) will the property constructor be called?

Comment (JamesNK) [reply src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs:463]: I don't understand the question.

The change is `GetRequiredService<IEnumerable<IHostedService>>()` can potentially be called multiple times. The registration methods for `IHostedService` sets them with a singleton lifetime: https://github.com/dotnet/runtime/blob/26990158d9725aa55d0af81f050a7a8e90c35f58/src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/ServiceCollectionHostedServiceExtensions.cs

That means even though `GetRequiredService<IEnumerable<IHostedService>>()` is called multiple times, the same instances are returned and there is no impact on behavior.

Comment (steveharter) [reply src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs:463]: I assume this always occurs after build(), so all services used for injection will already be added\registered, thus no side-affects.

**Review by steveharter — APPROVED** [review]

**Comment by JamesNK** [comment]
@steveharter I updated the PR to avoid resolving multiple times. It's unlikely, but possible, someone registers hosts with a non-singleton lifetime.

I wanted to write some unit tests that access the internal `Host` type, but I couldn't figure out how to get internals visible to to work in this repo.

**Review by steveharter — APPROVED** [review]

**Pull request merged**